### ISSUE TITLE
Questions about TCA: (1) normalization, (2) BCD vs HALS, (3) fit-transform

### ISSUE THREAD

**Issue by calebchiam** [ahwillia/tensortools#28]
Hi, I've read with a lot of interest your [paper on TCA](https://pubmed.ncbi.nlm.nih.gov/29887338/) and attempted to apply tensortools to my own work in NLP. I was wondering if you'd be willing to help answer some of my questions regarding TCA:

1. Should the feature columns of the data in the tensor be normalized prior to the TCA decomposition? (A normalization step was not mentioned in the paper, but my understanding is that the neural data in that setting were already on the same scale and thus did not require normalization.)

2. Do you have any advice for choosing between BCD and HALS for non-negative TCA? I've tried both and in some instances, there is no difference in the outputs, while in others there is a significant difference. (I also used the outputs in certain classification tasks and it was quite mixed as to whether the BCD or HALS output resulted in better predictions.)

3. I'm interested in applying the decomposition in a predictive setting. I understand that PCA is a lower-order analogue to TCA. [PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) can be fit() to a given matrix and has a repeatable transform() step that can be applied to new data. Is there an equivalent to this in TCA or would this be theoretically impossible? If the former, would it be possible to include this in tensortools as well?

Thanks very much!

**Comment by ahwillia** [comment]
Great questions.

1) Normalization will depend on the task. As a general rule of thumb, if any slice of the tensor is much "larger" than others, e.g. if `norm(X[:, i, :]) >> norm(X[:, j, :])`, then the model will be pre-disposed to fit slice `i` but ignore slice `j`.

2) No they both should perform similarly... Perhaps try both for your task and take whichever performs better. You should always run multiple optimization runs from different initial conditions anyway.

3) There isn't a really a clean analogue to the `transform` function in PCA unfortunately. In PCA you obtain orthogonal factors, so `U U^T` is a projection matrix (assuming here, the columns of `U` hold the factors). TCA gives you three sets of factor matrices `U`, `V`, `W`, but their columns are not an orthonormal basis. To make projection matrices you can compute `U inv(U^T U) U`, `V inv(V^T V) V`, `W inv(W^T W) W` for each of the three modes... In my experience, this generally doesn't provide much insight into the data.

If you are working in a predictive setting you could work with the factor matrices directly:

```python
model = tt.ncp_bcd(data, rank=R)
X = model.factors[2]  # if you are interested in predicting using features of axis=2
y = ...               # whatever your dependent variable is.

from sklearn.linear_model import Ridge  # or some other Regressor class
ridge = Ridge(alpha=1.0)
ridge.fit(X, y)
```
But instead of fitting two different models (first a tensor decomposition, then a regression model) it may be better for you to consider a unified tensor regression model... See e.g. http://www.jmlr.org/papers/volume18/16-362/16-362.pdf

Hope that helps!

**Comment by calebchiam** [comment]
Hi @ahwillia, thanks so much for the quick and detailed response! I have two quick follow-up questions:

1. Re: #3, I've tried working with the factor matrices directly for predictive tasks and they work reasonably well. One methodological concern I have however is that because of the lack of a separate transform() step, I don't see a way of rigorously evaluating how well the TCA method performs. This is because there is no true 'train-test split' (i.e. a way of having held-out data) when the factors themselves must first be obtained by decomposing both the train and test data that I have. Even with a tensor regression model, the same issue would still arise. Is my understanding correct or did I miss something?

(p.s. I'm attempting to construct the projection matrices you described for the purposes of creating such a transform() step. Thanks again for this!)

2. With PCA, there is an understanding that the components outputted are ordered according to 'importance' (i.e. the size of the eigenvalue). Does TCA have a similar guarantee? I've done several tests trying to confirm if the components are ordered by the norm, or decrease in reconstruction error, and such, but I've not been able to identify a consistent pattern in the ordering of the components.

Thank you again!

**Comment by ahwillia** [comment]
re / crossvalidation -- you can use the `mask` parameter in the `ncp_hals` function to implement splits into train / test sets. See figure 5 in the paper -- http://alexhwilliams.info/pdf/2018_tensor.pdf

re / ordering the components -- see discussion at https://github.com/ahwillia/tensortools/issues/17

**Comment by calebchiam** [comment]
Re: cross-validation, to clarify, by predictive setting, I actually mean a setup where if I was using PCA, I would: (1) PCA.fit_transform(train_data), (2) Train regression model on PCA components, (3) PCA.transform(test_data), (4) Evaluate performance of regression model on PCA components of test_data.

However, with TCA, I would not be able to replicate this setup because I would not be able to project the new test_data to the same 'component subspace' as obtained from the training step. The only way to do so would be to run the TCA decomposition on both train and test data, which poses a methodological issue. Does that make sense?
